src/CoreLib/distributedfunction.fs

```diff
@@ -343,6 +343,115 @@ type internal DistributedFunctionHolder(name:string, capacity:int, executor: (Gu
+                            chainedCTS.Dispose()
+                            semaphore.Release() |> ignore
+                        with 
+                        | ex -> 
+                            Logger.LogF( LogLevel.Warning, fun _ -> sprintf "ExecuteWithTimebudget, Functioon %s, exception in main loop of %A" name ex )
+                            observer.OnError( ex )
+                            chainedCTS.Dispose()
+                            semaphore.Release() |> ignore
+                    else
+                        try 
+                            if useTimeBudget <= elapse then 
+                                let ex = OperationCanceledException(sprintf "ExecuteWithTimebudget: Function %s, semaphore task is entered as cancelled" name)
+                                observer.OnError(ex)
+                            else
+                                let ex = TimeoutException( sprintf "ExecuteWithTimebudget: Function %s, semaphore task is entered in %d ms(budget %d ms)" name elapse timeBudget )
+                                observer.OnError(ex)
+                            chainedCTS.Dispose()
+                            semaphore.Release() |> ignore 
+                        with
+                        | ex -> 
+                            Logger.LogF( LogLevel.Warning, fun _ -> sprintf "ExecuteWithTimebudget, Functioon %s, exception while not enter, of %A" name ex )
+                            observer.OnError( ex )
+                            chainedCTS.Dispose()
+                let cancelledExecute (taskEnter:Task<bool>) = 
+                    try 
+                        chainedCTS.Dispose()
+                        semaphore.Release() |> ignore 
+                    with
+                    | ex -> 
+                        Logger.LogF( LogLevel.Warning, fun _ -> sprintf "ExecuteWithTimebudget, Functioon %s, exception when dispose of %A" name ex )
+                semaTask.ContinueWith( wrapperExecute, TaskContinuationOptions.None ) |> ignore 
+                semaTask.ContinueWith( cancelledExecute, TaskContinuationOptions.OnlyOnCanceled ) |> ignore
+            else
+                try 
+                    executor( jobID, useTimeBudget, input, chainedCTS.Token, observer )
+                    chainedCTS.Dispose()
+                with 
+                | ex -> 
+                    chainedCTS.Dispose()
+                    observer.OnError( ex )
+    override x.ToString() = sprintf "Distributed function %s:%s/%d" name (if Utils.IsNull semaphore then "Unknown" else (capacity-semaphore.CurrentCount).ToString() ) (capacity)
+    /// Cancel all jobs related to this distributed function. 
+    member x.Cancel() = 
+        // This process is responsible for the disposing routine
+        cts.Cancel()
+    /// Can we dispose this job holder?
+    member x.CanDispose() = 
+        // Wait for all job in semaphore to exit
+        if capacityControl then 
+            semaphore.CurrentCount = capacity 
```

Comment:
Distributed function calls a user provided function, which is not guaranteed to behave nicely. The user provided function itself can hang forever (meaning the executor(_) never returns). 

 

For caller, we can return a timeout exception. 

 

For callee (provider), I felt that there is no easy way to prevent malfunction of the provider. 

 

One possibility is that for those executor which implements cooperative cancellation, the operation can be cancelled if a system time out value is reached (say 600 sec). Even such timeout behavior is tricky to implement, as a system wide timeout value is risky to implement (it is highly possible that some rate operation will indeed need to take a long time to return). 

 

For executor which doesn’t implements cooperative cancellation, which will be the majority of the case for the distributed function, there is probably no way out (no nice way to cancel a function once it launches). 

 

 

Jin

 

 

From: Weirong Zhu [mailto:notifications@github.com] 
Sent: Saturday, October 17, 2015 4:50 PM
To: MSRCCS/Prajna <Prajna@noreply.github.com>
Cc: jinlmsft <jinlmsft@hotmail.com>
Subject: Re: [Prajna] Alternative Distributed function executor by SemaphoreSlim. (#79)

 

In src/CoreLib/distributedfunction.fs <https://github.com/MSRCCS/Prajna/pull/79#discussion_r42315429> :

> +                    executor( jobID, useTimeBudget, input, chainedCTS.Token, observer )
> +                    chainedCTS.Dispose()
> +                with 
> +                | ex -> 
> +                    chainedCTS.Dispose()
> +                    observer.OnError( ex )
> +    override x.ToString() = sprintf "Distributed function %s:%s/%d" name (if Utils.IsNull semaphore then "Unknown" else (capacity-semaphore.CurrentCount).ToString() ) (capacity)
> +    /// Cancel all jobs related to this distributed function. 
> +    member x.Cancel() = 
> +        // This process is responsible for the disposing routine
> +        cts.Cancel()
> +    /// Can we dispose this job holder?
> +    member x.CanDispose() = 
> +        // Wait for all job in semaphore to exit
> +        if capacityControl then 
> +            semaphore.CurrentCount = capacity 

I see. What's the mechanism to wait for these in-going work to exit? So we can safely call dispose?

—
Reply to this email directly or view it on GitHub <https://github.com/MSRCCS/Prajna/pull/79/files#r42315429> .  <https://github.com/notifications/beacon/AG7diAbzMIzyARd9W8xg7-6fuv6ymtwqks5o8tZAgaJpZM4GQqFq.gif>